src/pypyjs.js

```diff
@@ -0,0 +1,1113 @@
+      return this.stdin();
+    };
+
+    // Route stdout to an overridable method on the object.
+    // We buffer the output for efficiency.
+    const stdout = (byte) => {
+      const char = String.fromCharCode(byte);
+      stdoutBuffer.push(char);
+      if (char === '\n' || stdoutBuffer.length >= 128) {
+        this.stdout(stdoutBuffer.join(''));
+        stdoutBuffer = [];
+      }
+    };
+
+    // Route stderr to an overridable method on the object.
+    // We do not buffer stderr.
+    const stderr = (byte) => this.stderr(String.fromCharCode(byte));
+
+    // This is where execution will continue after loading
+    // the memory initialization data, if any.
+    let initializedResolve;
+    let initializedReject;
+    const initializedP = new Promise(function promise(_resolve, _reject) {
+      initializedResolve = _resolve;
+      initializedReject = _reject;
+    });
+
+    const dependenciesFulfilled = (_fs) => {
+      this.FS = _fs;
+
+      // Initialize the filesystem state.
+      try {
+        this.FS.init(stdin, stdout, stderr);
+        Module.FS_createPath('/', 'lib/pypyjs/lib_pypy', true, false);
+        Module.FS_createPath('/', 'lib/pypyjs/lib-python/2.7', true, false);
+        initializedResolve();
+      } catch (err) {
+        initializedReject(err);
+      }
+    };
+
+    // Begin fetching the metadata for available python modules.
+    // With luck these can download while we jank around compiling
+    // all of that javascript.
+    // XXX TODO: also load memory initializer this way.
+    const moduleDataP = this.fetch('modules/index.json');
+
+    pypyjs._vmBuilderPromise.then((vmBuilder) => {
+      const args = [
+        Module,
+        dependenciesFulfilled,
+        typeof require === 'undefined' ? undefined : require,
+        typeof module === 'undefined' ? undefined : module,
+        typeof __filename === 'undefined' ? undefined : __filename,
+        typeof _dirname === 'undefined' ? undefined : _dirname
+      ];
+
+      // This links the async-compiled module into our Module object.
+      vmBuilder.apply(null, args);
+      return initializedP;
+    }).then(() => {
+      // Continue with processing the downloaded module metadata.
+      return moduleDataP.then((xhr) => {
+        // Store the module index, and load any preload modules.
+        const modIndex = JSON.parse(xhr.responseText);
+        this._allModules = modIndex.modules;
+        if (modIndex.preload) {
+          Object.keys(modIndex.preload).forEach((name) => {
+            this._writeModuleFile(name, modIndex.preload[name]);
+          });
+        }
+
+        // It's finally safe to launch the VM.
+        Module.run();
+        Module._rpython_startup_code();
+        let pypy_home = Module.intArrayFromString('/lib/pypyjs/pypyjs.js');
+        pypy_home = Module.allocate(pypy_home, 'i8', Module.ALLOC_NORMAL);
+        Module._pypy_setup_home(pypy_home, 0);
+        Module._free(pypy_home);
+
+        const initCode = [
```

Comment:
sorry, I didn't notice this comment - no, no particular reason, other than the unreadability of multi-line strings in es5